Mathlib/RingTheory/Ideal/PrimeAvoidanceLemma.lean

```diff
@@ -0,0 +1,160 @@
+import Mathlib.Tactic.IntervalCases
+
+/-!
+# Prime Avoidance Lemma
+
+Let `R` be a commutative ring, `J` an ideal of `R`, `ℐ` be a finite collection of ideals of `R`
+such that ideals in `ℐ` are prime ideals except for perhaps at most two.
+
+If `J` is a subset of any of ideal in `ℐ`, then there is an `x ∈ R` such that `x ∈ J` but `x` is
+not in any of the ideals in `ℐ`.
+
+## TODO
+* graded version
+* the version where `R` contains an infinite field.
+
+## Reference
+[00DS](https://stacks.math.columbia.edu/tag/00DS)
+-/
+
+variable {R : Type _} [CommRing R]
+variable [DecidablePred fun I : Ideal R => I.IsPrime]
+
+open BigOperators
+open SetLike (coe_subset_coe)
+open Finset hiding not_subset
+open Set hiding mem_singleton mem_insert
+
+lemma Ideal.le_or_le_of_subset_union (J X Y : Ideal R) (H : (J : Set R) ⊆ X ∪ Y) :
+    J ≤ X ∨ J ≤ Y := by
+  by_contra rid
+  push_neg at rid
+  erw [not_subset_iff_exists_mem_not_mem, not_subset_iff_exists_mem_not_mem] at rid
+  rcases rid with ⟨⟨x, hx1, hx2⟩, ⟨y, hy1, hy2⟩⟩
+  rcases H (J.add_mem hx1 hy1) with h|h
+  · refine (H hy1).elim (fun h' => hx2 ?_) (fun h' => hy2 h')
+    convert X.sub_mem h h'; aesop
+  · refine (H hx1).elim (fun h' => hx2 h') (fun h' => hy2 ?_)
+    convert Y.sub_mem h h'; aesop
+
+/--
+Let `R` be a commutative ring, `J` an ideal of `R`, `ℐ` be a finite collection of ideals of `R`
+such that ideals in `ℐ` are prime ideals except for perhaps at most two.
+Then if `J` is a subset of the union of `ℐ`, `J` is already a subset of some ideal `I` in `ℐ`.
+-/
+theorem Ideal.le_of_subset_union_with_at_most_two_non_primes
+    (J : Ideal R)
+    (ℐ : Finset (Ideal R))
+    (number_of_non_prime : (ℐ.filter fun I => ¬ I.IsPrime).card ≤ 2)
+    (subset_union : (J : Set R) ⊆ ⋃ (I : ℐ), I) :
+    ∃ I, I ∈ ℐ ∧ J ≤ I := by
+  classical
+  induction' ℐ using Finset.strongInductionOn with ℐ ih
+  -- We perform a strong induction on `ℐ`, i.e. we assume that for any proper subset `𝒥` of `ℐ` with
+  -- at most two non-prime ideals, if `J` is a subset of the union of `𝒥`, then `I` is a subideal of
+  -- some ideal in `𝒥` already.
+
+  -- We can assume without lose of generality that `ℐ` has more than 2 ideals, for `ℐ` with fewer
+  -- ideals are easy cases.
+  by_cases card : ℐ.card ≤ 2
+  · replace card : ℐ.card = 0 ∨ ℐ.card = 1 ∨ ℐ.card = 2
+    · interval_cases ℐ.card <;> tauto
+    obtain card|card|card := card
+    · aesop
+    · obtain ⟨i, rfl⟩ := Finset.card_eq_one.mp card
+      exact ⟨i, mem_singleton_self _, fun x hx ↦ by aesop⟩
+    · obtain ⟨a, b, -, rfl⟩ := Finset.card_eq_two.mp card
+      simp only [mem_singleton, mem_insert, coe_subset_coe, exists_eq_or_imp, exists_eq_left,
+        iUnion_subtype, iUnion_iUnion_eq_or_left, iUnion_iUnion_eq_left] at subset_union ⊢
+      exact J.le_or_le_of_subset_union _ _ subset_union
+
+  -- We further assume that `J` is not a subset of any proper subset of `ℐ`, for otherwise, our
+  -- induction hypotheses implies the desired result already.
```

Comment:
When I reached the end reading the first time, I was a bit confused where the actual contradiction comes in to our proof. It's here:
```suggestion
  -- induction hypotheses implies the desired result already.
  -- We will show that this assumption in fact leads to a contradiction,
  -- since the goal is to produce an `I ≥ J`, such that `{I}` is a proper subset of `ℐ`.
```